pkg/util/kms.go

```diff
@@ -0,0 +1,277 @@
+package util
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/libopenstorage/secrets"
+	"github.com/libopenstorage/secrets/vault"
+	nbv1 "github.com/noobaa/noobaa-operator/v5/pkg/apis/noobaa/v1alpha1"
+	"github.com/noobaa/noobaa-operator/v5/pkg/bundle"
+	corev1 "k8s.io/api/core/v1"
+)
+
+///////////////////////////////////
+/////////// VAULT UTILS ///////////
+///////////////////////////////////
+const (
+	rootSecretPath          = "NOOBAA_ROOT_SECRET_PATH"
+	vaultCaCert             = "VAULT_CACERT"
+	vaultClientCert         = "VAULT_CLIENT_CERT"
+	vaultClientKey          = "VAULT_CLIENT_KEY"
+	VaultAddr               = "VAULT_ADDR"
+	vaultCaPath             = "VAULT_CAPATH"
+	VaultBackendPath        = "VAULT_BACKEND_PATH"
+	vaultToken              = "VAULT_TOKEN"
+	KmsProvider             = "KMS_PROVIDER"
+	KmsProviderVault        = "vault"
+	defaultVaultBackendPath = "secret/"
+)
+
+
+// VerifyExternalSecretsDeletion checks if noobaa is on un-installation process
+// if true, deletes secrets from external KMS
+func VerifyExternalSecretsDeletion(kms nbv1.KeyManagementServiceSpec, namespace string, uid string) error {
+
+	if len(kms.ConnectionDetails) == 0 {
+		log.Infof("deleting root key locally")
+		return nil
+	}
+
+	if !isVaultKMS(kms.ConnectionDetails[KmsProvider]) {
+		log.Errorf("Unsupported KMS provider %v", kms.ConnectionDetails[KmsProvider])
+		return fmt.Errorf("Unsupported KMS provider %v", kms.ConnectionDetails[KmsProvider])
+	}
+
+	c, err := InitVaultClient(kms.ConnectionDetails, kms.TokenSecretName, namespace)
+	if err != nil {
+		log.Errorf("deleting root key externally failed: init vault client: %v", err)
+		return err
+	}
+
+	secretPath := BuildExternalSecretPath(kms, uid)
+	err = DeleteSecret(c, secretPath)
+	if err != nil {
+		log.Errorf("deleting root key externally failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// InitVaultClient inits the secret store
+func InitVaultClient(config map[string]string, tokenSecretName string, namespace string) (secrets.Secrets, error) {
+
+	// create a type correct copy of the configuration
+	vaultConfig := make(map[string]interface{})
+	for k, v := range config {
+		vaultConfig[k] = v
+	}
+
+	// create TLS files out of secrets
+	// replace in vaultConfig secret names with local temp files paths
+	err := tlsConfig(vaultConfig, namespace)
+	if err != nil {
+		return nil, fmt.Errorf(`❌ Could not init vault tls config %q in namespace %q`, config, namespace)
+	}
+
+	// veify backend path, use default value if not set
+	if b, ok := config[VaultBackendPath]; !ok || b == "" {
+		log.Infof("KMS: using default backend path %v", defaultVaultBackendPath)
+		vaultConfig[VaultBackendPath] = defaultVaultBackendPath
```

Comment:
if I remember correctly, libopenstorage/secrets doing it by itself, I added it since I didn't use the library.
please check it and if it's not needed remove it

Reply:
Yep, you are correct 👍 [they do](https://github.com/libopenstorage/secrets/blob/a1d19aa9713a820d42fb3743ab8563d192d81ff9/vault/vault.go#L103-L106).